[BAHIR-110] Implement _changes API for non-streaming receiver

emlaver opened the pull request:
See [JIRA-110](https://issues.apache.org/jira/browse/BAHIR-110)

_What_

Add support for _changes API for non-streaming (data frames and SQL temp. views) receiver.

_How_
- New Cloudant config option `endpoint` for selecting `_all_docs` and `_changes` endpoint in Cloudant to Spark data frames and SQL temp tables
- Default is `_all_docs` endpoint for non-streaming receiver
- Base abstract config class that's extended by an all_docs class and _changes class
- JsonStoreConfigManager includes new 'cloudant.apiReceiver' config option for selecting _all_docs and _changes endpoint in Cloudant to Spark data frames and SQL temp tables
- Updated README with details for `cloudant.endpoint` option

_Testing_

- Added base class ClientSparkFunSuite for setting up, creating, and loading sample data from flat files to test databases.
- CloudantAllDocsDFSuite to test Spark data frames using the _all_docs endpoint.
- CloudantChangesDFSuite to test Spark data frames using the _changes endpoint.
- CloudantOptionSuite to verify Cloudant config options.
- CloudantSparkSQLSuite to test Spark SQL temp views.

Note: 27,378 lines added for the JSON files used in the testing suite.

ApacheBahir commented:
Can one of the admins verify this patch?

emlaver commented:
@lresende Environment variables for testing on Jenkins:
CLOUDANT_HOST (if not set, it will just do CLOUDANT_USER + `.cloudant.com`)
CLOUDANT_USER 
CLOUDANT_PASSWORD
CLOUDANT_PROTOCOL (default is https so don't need to set)

emlaver commented:
@yanglei99 Can you please review and add any comments?

ckadner commented on sql-cloudant/src/test/scala/org/apache/bahir/cloudant/TestUtils.scala:
```diff
@@ -0,0 +1,73 @@
+  )
+
+  // Set CouchDB/Cloudant host, username and password for local testing
+  private val host = System.getenv("DB_HOST")
```
@emlaver -- could you prefix the environment variables with **CLOUDANT_** ?

```
CLOUDANT_DB_HOST
CLOUDANT_DB_USER
CLOUDANT_DB_PASSWORD
CLOUDANT_DB_PROTOCOL
```

emlaver replied:
@ckadner I can do that or I can do `CLOUDANT_HOST`.  Eventually I'd like to have `COUCHDB_HOST` as well.  My preference is `CLOUDANT_HOST` and get rid of the DB.

emlaver replied:
Updated to `CLOUDANT_...` in 0256fc1.

emlaver commented:
@ckadner Updated tests to check if env. variables CLOUDANT_USER and CLOUDANT_PASSWORD exist.  If they do not exist, all tests will be skipped in d57cfe7

ckadner commented:
thanks @emlaver -- lets kick off a test build without the environment variables set

ckadner commented:
ok to test

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/65/

emlaver commented:
WIP: While running tests against databases with a size > 500 MB, `java.lang.OutOfMemoryError: Java heap space` error would occur (even when setting `--conf spark.driver.memory=10g`).  I believe this has to do with how the HTTP request is setup and called against the `_changes API` in [JsonStoreDataAccess.scala](https://github.com/apache/bahir/pull/45/files#diff-ab440bd537d48f7cf58cd9cf0ea143b1).
Good news is I've created a test that uses Spark streaming (using CloudantReceiver.java) to read all docs from a Cloudant database into a Spark DataFrame.  It should work for a SQL Temp Table.  I ran this test without any java heap errors against a database size of 1 GB, 1.8 GB, and 14.2 GB.

mayya-sharipova commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,61 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+Performance of `_changes` API is still better in most cases (even with single threaded support). 
+During several performance tests using 50 to 200 MB Cloudant databases, load time from Cloudant to Spark using `_changes` 
+feed was faster to complete every time compared to `_all_docs`.
+ 
```
with `_changes`, you can specify `selector` option using CQ format, thus significantly limit the number of documents to be loaded into Spark.

ckadner replied:
the code style guide (enforced by build) is 100 characters per line

mayya-sharipova commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantAllDocsConfig.scala:
```diff
@@ -0,0 +1,67 @@
+                           (username: String, password: String, partitions: Int,
+                            maxInPartition: Int, minInPartition: Int, requestTimeout: Long,
+                            bulkSize: Int, schemaSampleSize: Int,
+                            createDBOnSave: Boolean, apiReceiver: String, selector: String,
```
lines seem to be > 80 characters long

ckadner replied:
@mayya-sharipova -- code style guide calls for 100 character per line limit, which is enforced by the maven build's `verify` target

See: https://github.com/apache/bahir/blob/fd4c35fc9f7ebb57464d231cf5d66e7bc4096a1b/scalastyle-config.xml#L78

mayya-sharipova commented:
The `selector`  option doesn't work when  setting in config. For example, the below code will output all data without respecting `selector` settings.

```
spark = SparkSession\
    .builder\
    .appName("Cloudant Spark SQL Example in Python using dataframes")\
    .config("cloudant.host", ACC)\
    .config("cloudant.username", USER)\
    .config("cloudant.password", PASS)\
    .config("cloudant.apiReceiver", "_changes")\
    .config("selector", "{\"Movie_year\": {\"$eq\": 2008}}")\
    .getOrCreate()
```

mayya-sharipova commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,61 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+1. Results are partially ordered and may not be be presented in order in 
+which documents were updated.
+2. In case of shards' unavailability, you may see duplicate results (changes that have been seen already)
+3. Can use `selector` option to retrieve all revisions for docs
```
what do you mean by "all revisions" here?  I would say smth like this: " can use `selector` to filter docs to be loaded"

mayya-sharipova commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,61 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+2. In case of shards' unavailability, you may see duplicate results (changes that have been seen already)
+3. Can use `selector` option to retrieve all revisions for docs
+4. Only supports single threaded
+
```
You can add here that `_changes` support the real snapshot of the database representing it in a single point of time. While using `_all_docs` using partitions may not represent the true snapshot of the database, as some some docs may be added/deleted in the database between loading data into diff spark partitions.

mayya-sharipova commented:
The code using `option` parameters work as expected, and dbs are successfully loaded through `_changes`. But it seems to me that there could be better separation between classes `CloudantAllDocsConfig` and `CloudantChangesConfig`. For example, even though I am loading data using `_changes`, I have noticed there is still unnecessary calculations for the partitions going on. A part of it, is an extra unnecessary http request to calculate the number of `totalRows` in `JsonStoreRDD`: `https://ACCOUNT.cloudant.com/DB/_changes?limit=1&filter=_selector`

mayya-sharipova commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantAllDocsConfig.scala:
```diff
@@ -0,0 +1,67 @@
+  createDBOnSave, apiReceiver, selector, useQuery, queryLimit) {
+
+  override val defaultIndex: String = apiReceiver
+
```
Adding small comments for functions `getOneUrl`, `getUrl`, and `getSubSetUrl` describing how these functions are used would be a great help for understanding the code.

emlaver replied:
Added some details in 47064b7.

mayya-sharipova commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala:
```diff
@@ -0,0 +1,108 @@
+      }
+    }
+  }
+
```
It not clear what `getLastUrl` function is used for

emlaver replied:
Removed in 47064b7.

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/70/

mayya-sharipova commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+path| | Cloudant: as database name if database is not present
+schemaSampleSize|-1| the sample size used to discover the schema for this temp table. -1 scans all documents
+selector|all documents| a selector written in Cloudant Query syntax, specifying conditions for selecting documents when the `cloudant.apiReceiver` option is set to `_changes`. Only documents satisfying the selector's conditions will be retrieved from Cloudant and loaded into Spark.
+storageLevel|MEMORY_ONLY_SER| the storage level when persisting Spark data sets during load when `cloudant.apiReceiver` option equals `_changes`
```
We should also state here other options, or at least provide a link to other options.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -31,15 +31,14 @@ The `--packages` argument can also be used with `bin/spark-submit`.
 Submit a job in Python:
     
-    spark-submit  --master local[4] --jars <path to cloudant-spark.jar>  <path to python script> 
+    spark-submit  --master local[4] --packages org.apache.bahir:spark-sql-cloudant_2.11:2.2.0-SNAPSHOT  <path to python script> 
```
Should the example really point to a `SNAPSHOT` version?

emlaver replied:
Good point, this should probably be something like `--packages org.apache.bahir:spark-sql-cloudant_SCALA_VERSION:PACKAGE_VERSION`

emlaver replied:
Fixed in 5588c87.  There are existing examples on how to run this command under sections `Using SQL In Python` and `Using SQL In Scala`.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -31,15 +31,14 @@ The `--packages` argument can also be used with `bin/spark-submit`.
 Submit a job in Scala:
 
-	spark-submit --class "<your class>" --master local[4] --jars <path to cloudant-spark.jar> <path to your app jar>
+	spark-submit --class "<your class>" --master local[4] --packages org.apache.bahir:spark-sql-cloudant_2.11:2.2.0-SNAPSHOT <path to spark-sql-cloudant jar>
```
`SNAPSHOT` again?

emlaver replied:
Fixed in 5588c87.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
 
 Name | Default | Meaning
 --- |:---:| ---
+cloudant.apiReceiver|"_all_docs"| API endpoint for RelationProvider when loading or saving data from Cloudant to DataFrames or SQL temporary tables. Select between "_all_docs" or "_changes" endpoint.
```
I would say
>Cloudant API endpoint to use for

or
>Select between the Cloudant `_all_docs` or `_changes` endpoint

Probably also worth a "see below for details|notes|further explanation" or similar to help people find the content that would help them choose the appropriate endpoint.

emlaver replied:
Fixed in 5588c87.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
 bulkSize|200| the bulk save size
-schemaSampleSize| "-1" | the sample size for RDD schema discovery. 1 means we are using only first document for schema discovery; -1 means all documents; 0 will be treated as 1; any number N means min(N, total) docs 
-createDBOnSave|"false"| whether to create a new database during save operation. If false, a database should already exist. If true, a new database will be created. If true, and a database with a provided name already exists, an error will be raised. 
+schemaSampleSize|-1| the sample size for RDD schema discovery. 1 means we are using only first document for schema discovery; -1 means all documents; 0 will be treated as 1; any number N means min(N, total) docs 
```
using only **the** first document

emlaver replied:
Fixed in 5588c87.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+schemaSampleSize|-1| the sample size for RDD schema discovery. 1 means we are using only first document for schema discovery; -1 means all documents; 0 will be treated as 1; any number N means min(N, total) docs 
+createDBOnSave|false| whether to create a new database during save operation. If false, a database should already exist. If true, a new database will be created. If true, and a database with a provided name already exists, an error will be raised. 
+
+The `cloudant.apiReceiver` option allows for _changes or _all_docs API endpoint to be called while loading Cloudant data into Spark DataFrames or SQL Tables,
```
`allows` doesn't seem right here, maybe `sets` as in
>option sets the `_changes` or `_all_docs`

?

emlaver replied:
Fixed in 5588c87.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
+
+Performance of `_changes` API is still better in most cases (even with single threaded support). 
+During several performance tests using 200 MB to 15 GB Cloudant databases, load time from Cloudant to Spark using 
+`_changes` feed was faster to complete every time compared to `_all_docs`.
```
I think it is good to qualify the circumstances under which `_changes` performs better than `_all_docs` but this seems to me a bit of a teaser. I think this probably isn't the place to provide a complete testing report so perhaps a slightly more summarized version.

>Despite the restriction of single-threaded support the performance of the `_changes` API has been observed to be faster than `_all_docs` in testing with Cloudant databases up to 15 GB in size.

?

emlaver replied:
After our Slack discussion and @yanglei99 comments by e-mail, I've removed this in 5588c87.

ricellis commented on sql-cloudant/README.md:
```diff
@@ -52,39 +51,71 @@ Here each subsequent configuration overrides the previous one. Thus, configurati
-createDBOnSave|"false"| whether to create a new database during save operation. If false, a database should already exist. If true, a new database will be created. If true, and a database with a provided name already exists, an error will be raised. 
+createDBOnSave|false| whether to create a new database during save operation. If false, a database should already exist. If true, a new database will be created. If true, and a database with a provided name already exists, an error will be raised. 
+database| | Cloudant database name
+index| | Cloudant search index w/o the database name. only used for load data with less than or equal to 200 results.
```
I would say `Index queries are limited to returning 200 results so can only be used to load data with <= 200 results.`

emlaver replied:
Fixed in 5588c87.

emlaver replied:
@ricellis I removed the first part of the comment by accident.  In the documentation, I only found a limitation of 200 results for `search` indexes (no limits for Cloudant Query indexes).
I should change it to:
`Cloudant Search index without the database name.  Search index queries are limited to returning 200 results so can only be used to load data with <= 200 results.`

ricellis commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala:
```diff
@@ -0,0 +1,64 @@
+      selector
+    } else {
+      // Exclude design docs
+      "{ \"_id\": { \"$regex\": \"^(?!.*_design/)\" } }"
```
Does this regex really need a `.*`? Doesn't it make it exclude a document with a name like `my_design`? I guess the trailing slash protects against that because the `/` will appear in a design doc, but not a normal doc, but someone could create a doc with `/` I'm not sure if it would be encoded at the point of the regex check. Maybe more straightforward just to have `^(?!_design/)`

emlaver replied:
I don't think it needs the `.*`.  It's probably worth having adding a test case to verify this.

ricellis commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantConfig.scala:
```diff
@@ -30,81 +28,83 @@ import org.apache.bahir.cloudant.common._
-      if (limit == JsonStoreConfigManager.ALL_DOCS_LIMIT) {
+      val baseUrl = {
+        if (excludeDDoc) {
+          dbUrl + "/_all_docs?startkey=%22_design0/%22&include_docs=true"
```
See https://github.com/cloudant/java-cloudant/issues/344#issuecomment-276938689
This `startkey` makes the assumption that no document IDs will start with upper case letters. Possibly one for another issue.

emlaver replied:
I'll open a new JIRA issue for this.

ricellis commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantReceiver.scala:
```diff
@@ -16,23 +16,20 @@
 class CloudantReceiver(sparkConf: SparkConf, cloudantParams: Map[String, String])
     extends Receiver[String](StorageLevel.MEMORY_AND_DISK) {
-  lazy val config: CloudantConfig = {
+  lazy val config: CloudantChangesConfig = {
```
Would this class be better named `CloudantChangesReceiver`? Since it doesn't appear to be a general case.

emlaver replied:
From our conversation over Slack, let's not change the name of the `CloudantReceiver` class.

ricellis commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala:
```diff
@@ -98,29 +99,89 @@ class DefaultSource extends RelationProvider
+          ssc.start
+          // run streaming until all docs from continuous feed are received
+          ssc.awaitTermination
+          // ssc.stop(stopSparkContext = false, stopGracefully = false)
```
Commented out code?

emlaver replied:
Removed in 5588c87.

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/73/

mayya-sharipova commented on sql-cloudant/src/test/scala/org/apache/bahir/cloudant/CloudantSparkSQLSuite.scala:
```diff
@@ -0,0 +1,101 @@
+        .config("cloudant.host", TestUtils.getHost)
+        .config("cloudant.username", TestUtils.getUsername)
+        .config("cloudant.password", TestUtils.getPassword)
+        .config("cloudant.apiReceiver", apiReceiver)
```
@emlaver is it `apiReceiver` or `endpoint`?

emlaver replied:
@mayya-sharipova It's `endpoint`, looks like I missed this test suite.  Thanks for catching that.

emlaver replied:
Fixed in 9f8bbc5.

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/74/

mayya-sharipova commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala:
```diff
@@ -98,29 +99,81 @@ class DefaultSource extends RelationProvider
+
+          logger.info("Loading data from Cloudant using "
+            + config.asInstanceOf[CloudantChangesConfig].getContinuousChangesUrl)
+
```
@emlaver  Here while trying to load a db, I am getting a message "Loading data from Cloudant using https://XXXX.cloudant.com/n_airportcodemapping/_changes?include_docs=true&feed=continuous&heartbeat=3000"

We should NOT load data into Spark SQL using `continuous` feed (which for constantly updating database may never be over).  The whole point of loading a db into Spark SQL is to load a snapshot of a db at a particular point of time.  Use `normal` feed here.

emlaver replied:
@mayya-sharipova Updates and fixes in c2c7fc4.

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/75/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/76/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/77/

ckadner commented:
@emlaver -- the most recent build ([77](http://169.45.79.58:8080/job/bahir_spark_pr_builder/77/)) ran with user/pwd env vars set ...

```
Discovery completed in 185 milliseconds.
Run starting. Expected test count is: 21
CloudantOptionSuite:
- invalid api receiver option throws an error message
- empty username option throws an error message
- empty password option throws an error message
- empty databaseName throws an error message
ClientSparkFunSuite:
CloudantChangesDFSuite:
- load and save data from Cloudant database
- load and count data from Cloudant search index
- load data and count rows in filtered dataframe
- save filtered dataframe to database
- save dataframe to database using createDBOnSave=true option
- load and count data from view
- load data from view with MapReduce function
- load data and verify total count of selector, filter, and view option
CloudantSparkSQLSuite:
- verify results from temp view of database n_airportcodemapping
- verify results from temp view of index in n_flight
CloudantAllDocsDFSuite:
- load and save data from Cloudant database
- load and count data from Cloudant search index
- load data and count rows in filtered dataframe
- save filtered dataframe to database
- save dataframe to database using createDBOnSave=true option
- load and count data from view
- load data from view with MapReduce function

Run completed in 2 minutes, 58 seconds.
Total number of tests run: 21
Suites: completed 6, aborted 0
Tests: succeeded 21, failed 0, canceled 0, ignored 0, pending 0

All tests passed.
```

emlaver commented:
Stability tests update: successfully loaded Cloudant docs into Spark yesterday with the following database sizes: 15GB (time: 8 1/2 mins), 20GB (17 mins), 46GB (25 mins), and 75GB (48 1/2 mins).
@mayya-sharipova @yanglei99 @ricellis

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/78/

mayya-sharipova commented:
@emlaver 
I am getting the following unexpected behaviour:
I have a database with 13 docs and 1 deleted doc. When displaying `df.count`, I am getting `14` which is incorrect.  When displaying a dataframe, I am getting the last record of NULL.
+--------+---+--------------------+-----------+
|_deleted|_id|                _rev|airportName|
+--------+---+--------------------+-----------+
|    null|DEL|1-67f14f8891a9f32...|      Delhi|
|    null|JFK|1-ee8206c8e56a114...|   New York|
|    null|SVO|1-7d18769b68f6099...|     Moscow|
|    null|FRA|1-f358b62b0499340...|  Frankfurt|
|    null|HKG|1-b040e40df5d0080...|  Hong Kong|
|    null|CDG|1-8c51e401185272e...|      Paris|
|    null|FCO|1-89431c8db8aa8e4...|       Rome|
|    null|NRT|1-dce312ac1414110...|      Tokyo|
|    null|LHR|1-303c622ad8380c9...|     London|
|    null|BOM|2-a3f39a0741938c4...|    Mumbaii|
|    null|YUL|1-19a9fe9cace23ec...|   Montreal|
|    null|IKA|1-3dea74452ca86af...|     Tehran|
|    null|SIN|1-67037272289432e...|  Singapore|
|    true|SYD|2-1cc4f2c62db144a...|       null|
+--------+---+--------------------+-----------+

We should NOT load into dataframe any deleted documents. A user may have thousands or millions of deleted documents. We should load only undeleted docs, and a dataframe should NOT have a column `"_deleted"`.

________
Another error:
Occasionally, when running `CloudantDF.py` example, get an error:
File "/Cloudant/bahir/sql-cloudant/examples/python/CloudantDF.py", line 45, in <module>
    df.filter(df.airportName >= 'Moscow').select("_id",'airportName').show()
  File "..spark-2.2.0-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/sql/dataframe.py", line 1020, in __getattr__
AttributeError: 'DataFrame' object has no attribute 'airportName' 

For this PR, we can disregard this error and investigate further in following PRs.
_____________

ApacheBahir commented:
Build failed, see build log for details

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/79/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/80/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/81/

mayya-sharipova commented on sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala:
```diff
@@ -0,0 +1,100 @@
+                count += 1
+              }
+            }
+          } else if (count >= limit) {
```
should we check `(count >= limit)` before line 72 for all objects, not only empty ones?

emlaver replied:
Fixed in 8b40e38.

mayya-sharipova commented:
+1 from my side after addressing small comments! Great work and very necessary!  Please squash all commits before merging!!!

emlaver commented:
@mayya-sharipova Thank you for all the time you've spent reviewing this work.
The storageLevel option was not working as expected and also was not in the correct section in the README.  I’ve renamed the option to `cloudant.storageLevel` (as this option only works when _changes endpoint option is set), updated the README, and successfully tested and verified that the option is working using spark-submit and SparkSession’s config in 052abb4.  I've run the Scala tests and examples and they all passed.  Could you please review and approve these changes?

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/86/

emlaver commented:
@ckadner Do you have any additional comments? I'll rebase and squash the commits once @mayya-sharipova reviews the last commit.

ckadner commented:
Thanks @emlaver for this PR and @mayya-sharipova and @ricellis and @yanglei99 for your thorough review!

**LGTM**

emlaver commented:
@ckadner @ricellis @mayya-sharipova @yanglei99 Thanks for all your comments and time spent reviewing this large PR!

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/87/

ApacheBahir commented:
Build failed, see build log for details

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/88/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/89/

ApacheBahir commented:
Build successful

ApacheBahir commented:
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/90/
